hpx/runtime/components/server/migration_support.hpp

```diff
@@ -45,15 +78,39 @@ namespace hpx { namespace components
         void pin()
         {
             boost::lock_guard<mutex_type> l(mtx_);
-            ++pin_count_;
+            if (pin_count_ != ~0x0u)
+                ++pin_count_;
         }
         void unpin()
         {
-            boost::lock_guard<mutex_type> l(mtx_);
+            typename mutex_type::scoped_lock l(mtx_);
             HPX_ASSERT(pin_count_ != 0);
             if (pin_count_ != ~0x0u)
-                --pin_count_;
+            {
+                if (--pin_count_ == 0)
+                {
+                    // trigger pending migration if this was the last unpin and
+                    // a migration operation is pending
+                    if (trigger_migration_.valid() && migration_target_locality_)
+                    {
+                        migration_target_locality_ = naming::invalid_id;
+
+                        hpx::future<void> f =
+                            agas::mark_as_migrated(this->gid_,
+                                []() -> hpx::future<void>
+                                {
+                                    return make_ready_future();
+                                });
+
+                        trigger_migration_.set_value();
+
+                        l.unlock();     // avoid holding lock during suspension
```

Comment:
This unlock might be problematic. At this point, it is not guaranteed that the object has been marked as migrated in AGAS and actions might still be scheduled on to be migrated objects.